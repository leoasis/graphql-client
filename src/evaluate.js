import vm from "vm";
import m from "module";
import * as babel from "babel-core";

export default function evaluate(code) {
  const exported = {};
  const mod = vm.runInNewContext(
    m.wrap(
      babel.transform(code, {
        presets: ["env"],
        plugins: ["transform-object-rest-spread"]
      }).code
    )
  );
  mod(exported);
  return exported;
}
