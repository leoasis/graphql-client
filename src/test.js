import gql from "graphql-tag";
import Compiler from ".";
import evaluate from "./evaluate";

test("basic read and write query", () => {
  const compiler = new Compiler("../schema.graphql");

  const data = {
    hero: {
      id: "foo",
      __typename: "Droid",
      name: "R2D2"
    }
  };

  const code = compiler.compile(`
  query Foo {
    hero {
      id
      __typename
      name
    }
  }`);

  expect(code).toMatchSnapshot();

  const { writeQuery, readQuery } = evaluate(code);

  const written = writeQuery(data, {});
  expect(written).toEqual({
    ROOT_QUERY: {
      hero: "Droid:foo"
    },
    "Droid:foo": {
      __typename: "Droid",
      id: "foo",
      name: "R2D2"
    }
  });

  const read = readQuery(written);
  expect(read).toEqual(data);
});
test("more complex read and write query", () => {
  const compiler = new Compiler("../schema.graphql");

  const data = {
    hero: {
      __typename: "Droid",
      id: "2001",
      name: "R2-D2",
      friends: [
        {
          __typename: "Human",
          id: "1000",
          name: "Luke Skywalker",
          appearsIn: ["NEWHOPE", "EMPIRE", "JEDI"]
        },
        {
          __typename: "Human",
          id: "1002",
          name: "Han Solo",
          appearsIn: ["NEWHOPE", "EMPIRE", "JEDI"]
        },
        {
          __typename: "Human",
          id: "1003",
          name: "Leia Organa",
          appearsIn: ["NEWHOPE", "EMPIRE", "JEDI"]
        }
      ]
    }
  };

  const code = compiler.compile(`
  query Foo {
    hero {
      __typename
      id
      name
      friends {
        __typename
        id
        name
        appearsIn
      }
    }
  }
  `);

  expect(code).toMatchSnapshot();

  const { writeQuery, readQuery } = evaluate(code);

  const written = writeQuery(data, {});
  expect(written).toEqual({
    ROOT_QUERY: {
      hero: "Droid:2001"
    },
    "Droid:2001": {
      __typename: "Droid",
      id: "2001",
      name: "R2-D2",
      friends: ["Human:1000", "Human:1002", "Human:1003"]
    },
    "Human:1000": {
      __typename: "Human",
      id: "1000",
      name: "Luke Skywalker",
      appearsIn: ["NEWHOPE", "EMPIRE", "JEDI"]
    },
    "Human:1002": {
      __typename: "Human",
      id: "1002",
      name: "Han Solo",
      appearsIn: ["NEWHOPE", "EMPIRE", "JEDI"]
    },
    "Human:1003": {
      __typename: "Human",
      id: "1003",
      name: "Leia Organa",
      appearsIn: ["NEWHOPE", "EMPIRE", "JEDI"]
    }
  });

  const read = readQuery(written);
  expect(read).toEqual(data);
});

test("nested lists", () => {
  const compiler = new Compiler("../schema.graphql");

  const data = {
    hero: {
      __typename: "Droid",
      id: "2001",
      name: "R2-D2",
      friends: [
        {
          __typename: "Human",
          id: "1000",
          name: "Luke Skywalker",
          appearsIn: ["NEWHOPE", "EMPIRE", "JEDI"],
          friends: [
            {
              __typename: "Droid",
              id: "2001",
              name: "R2-D2"
            },
            {
              __typename: "Human",
              id: "1002",
              name: "Han Solo"
            }
          ]
        },
        {
          __typename: "Human",
          id: "1002",
          name: "Han Solo",
          appearsIn: ["NEWHOPE", "EMPIRE", "JEDI"],
          friends: []
        },
        {
          __typename: "Human",
          id: "1003",
          name: "Leia Organa",
          appearsIn: ["NEWHOPE", "EMPIRE", "JEDI"],
          friends: [
            {
              __typename: "Droid",
              id: "2001",
              name: "R2-D2"
            },
            {
              __typename: "Human",
              id: "1002",
              name: "Han Solo"
            },
            {
              __typename: "Human",
              id: "1000",
              name: "Luke Skywalker"
            }
          ]
        }
      ]
    }
  };

  const code = compiler.compile(`
  query Foo {
    hero {
      __typename
      id
      name
      friends {
        __typename
        id
        name
        appearsIn
        friends {
          __typename
          id
          name
        }
      }
    }
  }
  `);

  expect(code).toMatchSnapshot();

  const { writeQuery, readQuery } = evaluate(code);

  // const written = writeQuery(data, {});

  // Better write query implementation!!!
  // Change code to generate this:
  function writeQuery2(data, cache) {
    let v0, v1;
    const entities = [];
    const root = {
      hero: ((v0 = data.hero),
      (v1 = v0.__typename + ":" + v0.id),
      entities.push([
        v1,
        {
          __typename: v0.__typename,
          id: v0.id,
          name: v0.name,
          friends: v0.friends.map(item => {
            const id = item.__typename + ":" + item.id;
            entities.push([
              id,
              {
                __typename: item.__typename,
                id: item.id,
                name: item.name,
                appearsIn: item.appearsIn,
                friends: item.friends.map(item => {
                  const id = item.__typename + ":" + item.id;
                  entities.push([
                    id,
                    {
                      __typename: item.__typename,
                      id: item.id,
                      name: item.name
                    }
                  ]);
                  return id;
                })
              }
            ]);
            return id;
          })
        }
      ]),
      v1)
    };

    cache = {
      ...cache,
      ROOT_QUERY: root
    };

    entities.forEach(([id, entity]) => {
      cache[id] = { ...cache[id], ...entity };
    });

    return cache;
  }
  const written = writeQuery2(data, {});

  expect(written).toEqual({
    ROOT_QUERY: {
      hero: "Droid:2001"
    },
    "Droid:2001": {
      __typename: "Droid",
      id: "2001",
      name: "R2-D2",
      friends: ["Human:1000", "Human:1002", "Human:1003"]
    },
    "Human:1000": {
      __typename: "Human",
      id: "1000",
      name: "Luke Skywalker",
      appearsIn: ["NEWHOPE", "EMPIRE", "JEDI"],
      friends: ["Droid:2001", "Human:1002"]
    },
    "Human:1002": {
      __typename: "Human",
      id: "1002",
      name: "Han Solo",
      appearsIn: ["NEWHOPE", "EMPIRE", "JEDI"],
      friends: []
    },
    "Human:1003": {
      __typename: "Human",
      id: "1003",
      name: "Leia Organa",
      appearsIn: ["NEWHOPE", "EMPIRE", "JEDI"],
      friends: ["Droid:2001", "Human:1002", "Human:1000"]
    }
  });

  const read = readQuery(written);
  expect(read).toEqual(data);
});
