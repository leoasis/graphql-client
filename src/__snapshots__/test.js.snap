// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic read and write query 1`] = `
"export function writeQuery(data, cache) {
  let v0, v1;
  return {
    ...cache,
    \\"ROOT_QUERY\\": {
      \\"hero\\": (v0 = data.hero, v1 = v0.__typename + \\":\\" + v0.id, v1)
    },
    [v1]: {
      \\"id\\": v0.id,
      \\"__typename\\": v0.__typename,
      \\"name\\": v0.name
    }
  };
}
export function readQuery(cache) {
  const root = cache.ROOT_QUERY;
  let v0, v1;
  return {
    \\"hero\\": (v0 = root.hero, v1 = cache[v0], {
      \\"id\\": v1.id,
      \\"__typename\\": v1.__typename,
      \\"name\\": v1.name
    })
  };
}"
`;

exports[`more complex read and write query 1`] = `
"export function writeQuery(data, cache) {
  let v0, v1, v2;
  return {
    ...cache,
    \\"ROOT_QUERY\\": {
      \\"hero\\": (v0 = data.hero, v1 = v0.__typename + \\":\\" + v0.id, v1)
    },
    [v1]: {
      \\"__typename\\": v0.__typename,
      \\"id\\": v0.id,
      \\"name\\": v0.name,
      \\"friends\\": (v2 = v0.friends, v2.map(item => item.__typename + \\":\\" + item.id))
    },
    ...v2.reduce((obj, item) => {
      obj[item.__typename + \\":\\" + item.id] = item;
      return obj;
    }, {})
  };
}
export function readQuery(cache) {
  const root = cache.ROOT_QUERY;
  let v0, v1, v2;
  return {
    \\"hero\\": (v0 = root.hero, v1 = cache[v0], {
      \\"__typename\\": v1.__typename,
      \\"id\\": v1.id,
      \\"name\\": v1.name,
      \\"friends\\": (v2 = v1.friends, v2.map(item => cache[item]))
    })
  };
}"
`;

exports[`nested lists 1`] = `
"export function writeQuery(data, cache) {
  let v0, v1, v2;
  return {
    ...cache,
    \\"ROOT_QUERY\\": {
      \\"hero\\": (v0 = data.hero, v1 = v0.__typename + \\":\\" + v0.id, v1)
    },
    [v1]: {
      \\"__typename\\": v0.__typename,
      \\"id\\": v0.id,
      \\"name\\": v0.name,
      \\"friends\\": (v2 = v0.friends, v2.map(item => item.__typename + \\":\\" + item.id))
    },
    ...v2.reduce((obj, item) => {
      obj[item.__typename + \\":\\" + item.id] = item;
      return obj;
    }, {})
  };
}
export function readQuery(cache) {
  const root = cache.ROOT_QUERY;
  let v0, v1, v2;
  return {
    \\"hero\\": (v0 = root.hero, v1 = cache[v0], {
      \\"__typename\\": v1.__typename,
      \\"id\\": v1.id,
      \\"name\\": v1.name,
      \\"friends\\": (v2 = v1.friends, v2.map(item => cache[item]))
    })
  };
}"
`;
